jena-arq/src/main/java/org/apache/jena/sparql/core/mem/DatasetGraphInMemory.java

```diff
@@ -0,0 +1,310 @@
+	}
+
+	protected void isInTransaction(final boolean b) {
+		isInTransaction.set(b);
+	}
+
+	private final ThreadLocal<ReadWrite> transactionType = withInitial(() -> null);
+
+	/**
+	 * @return the type of transaction in progress
+	 */
+	public ReadWrite transactionType() {
+		return transactionType.get();
+	}
+
+	protected void transactionType(final ReadWrite readWrite) {
+		transactionType.set(readWrite);
+	}
+
+	private final QuadTable quadsIndex;
+
+	private QuadTable quadsIndex() {
+		return quadsIndex;
+	}
+
+	private final TripleTable defaultGraph;
+
+	private TripleTable defaultGraph() {
+		return defaultGraph;
+	}
+
+	@Override
+	public Lock getLock() {
+		return writeLock();
+	}
+
+	/**
+	 * Default constructor.
+	 */
+	public DatasetGraphInMemory() {
+		this(new HexTable(), new TriTable());
+	}
+
+	/**
+	 * @param i a table in which to store quads
+	 * @param t a table in which to store triples
+	 */
+	public DatasetGraphInMemory(final QuadTable i, final TripleTable t) {
+		this.quadsIndex = i;
+		this.defaultGraph = t;
+	}
+
+	@Override
+	public void begin(final ReadWrite readWrite) {
+		if (isInTransaction()) throw new JenaTransactionException("Transactions cannot be nested!");
+		transactionType(readWrite);
+		isInTransaction(true);
+		getLock().enterCriticalSection(readWrite.equals(READ)); // get the dataset write lock, if needed.
+		commitLock().readLock().lock(); // if a commit is proceeding, wait so that we see a coherent index state
+		try {
+			quadsIndex().begin(readWrite);
+		} finally {
+			commitLock().readLock().unlock();
+		}
+	}
+
+	@Override
+	public void commit() {
+		if (!isInTransaction()) throw new JenaTransactionException("Tried to commit outside a transaction!");
+		commitLock().writeLock().lock();
+		try {
+			quadsIndex().commit();
+			defaultGraph().commit();
+		} finally {
+			commitLock().writeLock().unlock();
+		}
+		end();
+	}
+
+	@Override
+	public void abort() {
+		if (!isInTransaction()) throw new JenaTransactionException("Tried to abort outside a transaction!");
+		end();
+	}
+
+	@Override
+	public void close() {
+		if (isInTransaction()) abort();
+
+	}
+
+	@Override
+	public void end() {
+		quadsIndex().end();
+		defaultGraph().end();
+		isInTransaction(false);
+		transactionType(null);
+		getLock().leaveCriticalSection();
+	}
+
+	private <T> Iterator<T> access(final Supplier<Iterator<T>> source) {
+		if (!isInTransaction()) {
+			begin(READ);
+			try {
+				return source.get();
+			} finally {
+				end();
+			}
+		}
+		return source.get();
+	}
+
+	@Override
+	public Iterator<Node> listGraphNodes() {
+		return access(() -> quadsIndex().listGraphNodes().iterator());
+	}
+
+	private Iterator<Quad> quadsFinder(final Node g, final Node s, final Node p, final Node o) {
+		if (isUnionGraph(g)) { // union graph is the merge of named graphs
+			final Set<Triple> seen = new HashSet<>();
+			return quadsIndex().find(ANY, s, p, o).filter(q -> !q.isDefaultGraph() && seen.add(q.asTriple()))
+					.iterator();
+		}
```

Comment:
1. I'm not sure what you mean by "stream this". It _is_ streaming. Unless you mean to avoid the buildup of state in the `Set`, in exchange for your suggestion about adjacency? In that case, see below.
2. This is the `DatasetGraph`. It doesn't know anything about SPOG or any other particular kind of index. It only knows about `QuadTable`s and `TripleTable`s, which may happen (do, in the default case) to be implemented with maps, but could just as well be implemented with some other technique that wouldn't produce adjacency as an artifact. (E.g. @Claudenw 's idea about Bloom filters, or your suggestion about "mixed structures".)